cranelift/assembler-x64/src/reg.rs

```diff
@@ -0,0 +1,247 @@
+//! Pure register operands; see [`Gpr`].
+
+use crate::AsReg;
+use crate::{fuzz::FuzzReg, rex::RexFlags};
+use arbitrary::Arbitrary;
+
+/// A general purpose x64 register (e.g., `%rax`).
+///
+/// This container wraps true register type `R` to allow users to specify their
+/// own; by default this will use `u8`.
+#[derive(Clone, Copy, Debug)]
+pub struct Gpr<R: AsReg = u8>(pub(crate) R);
+
+impl<R: AsReg> Gpr<R> {
+    /// Create a [`Gpr`] that may be real (immediately emit-able in machine
+    /// code) or virtual (waiting for register allocation).
+    pub fn new(reg: R) -> Self {
+        Self(reg)
+    }
+
+    /// Return the register's hardware encoding; the underlying type `R` _must_
+    /// be a real register at this point.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the register is not a valid x64 register.
+    pub fn enc(&self) -> u8 {
+        let enc = self.0.enc();
+        assert!(enc < 16, "invalid register: {enc}");
+        enc
+    }
+
+    /// Return the register name at the given `size`.
+    pub fn to_string(&self, size: Size) -> &str {
+        enc::to_string(self.enc(), size)
+    }
+
+    /// Proxy on the 8-bit REX flag emission; helpful for simplifying generated
+    /// code.
+    pub(crate) fn always_emit_if_8bit_needed(&self, rex: &mut RexFlags) {
+        rex.always_emit_if_8bit_needed(self.enc());
+    }
+}
+
+impl<R: AsReg> AsRef<R> for Gpr<R> {
+    fn as_ref(&self) -> &R {
+        &self.0
+    }
+}
+
+impl<R: AsReg> AsMut<R> for Gpr<R> {
+    fn as_mut(&mut self) -> &mut R {
+        &mut self.0
+    }
+}
+
+impl<'a, R: AsReg> Arbitrary<'a> for Gpr<R> {
+    fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+        let reg = FuzzReg::arbitrary(u)?;
+        Ok(Self(R::new(reg.enc())))
+    }
+}
+
+/// A single x64 register encoding can access a different number of bits.
+#[derive(Copy, Clone, Debug)]
+pub enum Size {
+    /// An 8-bit access.
+    Byte,
+    /// A 16-bit access.
+    Word,
+    /// A 32-bit access.
+    Doubleword,
+    /// A 64-bit access.
+    Quadword,
+}
+
+/// Like [`Gpr`], but with `%rsp` disallowed.
+///
+/// This is due to avoid special cases of REX encodings, see Intel SDM Vol. 2A,
+/// table 2-5.
+#[derive(Clone, Copy, Debug)]
+pub struct MinusRsp<R: AsReg>(R);
```

Comment:
`MinusRsp` evokes some sort of subtraction operation in my mind; maybe `NonRspGpr` or something like that?